xsnap/documentation/Compartments.md

```diff
@@ -0,0 +1,618 @@
+
+#### module specifiers
+
+When the `moduleMapHook` function returns a module specifier, the module specifier is used by the parent compartment to load and initialize a new module into the compartment.
+
+	const uri = import.meta.uri;
+	const foo = uri.slice(0, uri.lastIndexOf("/") + 1) + "moduleMap-fixture.js";
+	
+	function moduleMapHook(specifier) {
+		if ((specifier == "foo") || (specifier == "bar"))
+			return foo;
+	}
+	
+	const c1 = new Compartment({}, {}, { moduleMapHook });
+	const fooNS1 = await c1.import("foo");
+	const barNS1 = await c1.import("bar");
+	
+	const c2 = new Compartment({}, {}, { moduleMapHook });
+	const fooNS2 = await c2.import("foo");
+	const barNS2 = await c2.import("bar");
+	
+	print(fooNS1.default()); // 0;
+	print(barNS1.default()); // 0;
+	print(fooNS2.default()); // 0;
+	print(barNS2.default()); // 0;
+
+### options.loadHook(specifier)
+
+The `loadHook` option is an asynchronous function that takes a module specifier and returns a promise to a module descriptor.
+
+The `loadHook` function is only called directly or indirectly by `Compartment.prototype.import` if the `moduleMap` has no property for the specifier and if the `moduleMapHook` returns `undefined` for the specifier.
+
+The `loadHook` function is useful if the module descriptor is unavailable when constructing the compartment, or to create a module descriptor dynamically. 
+
+	const c = new Compartment({}, {}, {
+		async loadHook(specifier) {
+			return { source:`export default import.meta.uri`, meta: { uri: specifier } };
+		}
+	})
+	const nsa = await c.import("a");
+	print(nsa.default); // a
+	const nsb = await c.import("b");
+	print(nsb.default); // b
+
+The default `loadHook` function throws a `TypeError`.
+
+### options.loadNowHook(specifier)
+
+The `loadNowHook` option is a function that takes a module specifier and returns a module descriptor.
+
+The `loadNowHook` function is only called directly or indirectly by `Compartment.prototype.importNow` if the `moduleMap` has no property for the specifier and if the `moduleMapHook` returns `undefined` for the specifier.
+
+The `loadNowHook ` function is useful if the module descriptor is unavailable when constructing the compartment, or to create a module descriptor dynamically. 
+
+	const c = new Compartment({}, {}, {
+		loadNowHook(specifier) {
+			return { source:`export default import.meta.uri`, meta: { uri: specifier } };
+		}
+	})
+	const nsa = c.importNow("a");
+	print(nsa.default); // a
+	const nsb = c.importNow("b");
+	print(nsb.default); // b
+
+The default `loadNowHook` function throws a `TypeError`.
+
+### options.resolveHook(importSpecifier, referrerSpecifier)
+
+The `resolveHook` option is a function that takes an import specifier and a referrer specifier and returns a module specifier.
+
+Typically the `resolveHook` function combine a relative path and an absolute path or uri into an absolute path or uri. But the `resolveHook` function can build arbitrary specifiers.
```

Comment:
Suggestion:

To emulate Node.js and the web, full specifiers including referrer specifiers would always be a fully qualified URL, identical to the `import.meta.url` and import specifiers are either fully qualified URL’s, relative specifiers (starting with `./`, or absolute specifiers (other non-URL forms), and the `resolveHook` would be relegated to the synchronous portion of module specifier resolution that can be performed synchronously. The `loadHook` then would necessarily take on the responsibility of following an import map, or package graph, HTTP redirects, or symbolic links the physical location of the module.

A full module specifier in XS and Endo is not a fully qualified URL since modules in such systems are location-independent. In Endo, a full specifier may be either a relative path starting with `./` or an absolute specifier (any other specifier). Relative specifiers indicate compartment-internal linkage and absolute specifiers indicate linkage to an external compartment. As with the web and Node.js, the `loadHook` may follow this logical specifier to a different physical location.

Regardless, there is no default `resolveHook` and resolve hooks by design afford a wide variety of module specifier languages.